docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+effects simultaneously. And that's where it gets difficult: as it turns out,
+combining effects is a huge design space, and there are many conflicting
+approaches (concrete transformer stacks, mtl-style classes, extensible effects,
+and what not).
+
+What further complicates matters is that the choice of the approach to monadic
+effects will affect the entire codebase in drastic ways, since the effects are
+so ubiqutous. This means that it's impossible to quickly try each possible
+approach, one must rewrite half the application to do so.
+
+History
+----------
+
+In Cardano SL we've used multiple approaches to effects, none of them
+satisfactory. The goal of this document is to provide a reference point for
+discussion about further direction of travel.
+
+The Stone Age
+~~~~~~~~~~~~~~
+
+The first approach to effects that was used in Cardano SL was simple: for each
+effect, create a class with operations and a monad transformer that implements
+these operations::
+
+    class MonadFoo m where
+        ...
+
+    data FooT m a = FooT { runFooT :: ... }
+
+    instance MonadFoo (FooT m)
+    instance MonadFoo m => MonadFoo (ReaderT r m)
+    instance MonadFoo m => MonadFoo (StateT s m)
+    instance MonadFoo m => MonadFoo (ExceptT e m)
+    ...
+
+    class MonadBar m where
+        ...
+
+    data BarT m a = BarT { runBarT :: ... }
+
+    instance MonadBar (BarT m)
+    instance MonadBar m => MonadBar (ReaderT r m)
+    instance MonadBar m => MonadBar (StateT s m)
+    instance MonadBar m => MonadBar (ExceptT e m)
+    ...
+
+This approach mirrors the one of ``mtl``. However, ``mtl`` is concerned with
+general purpose effects, of which there are only a few (under 10, even if we
+consider more exotic ones like reverse state or logic-t), whereas
+domain-specific effects are numerous. Besides, general purpose effects don't pop
+up left and right, whereas domain-specific ones might be created/deleted
+depending on how the application evolves. All in all, this made mirroring
+``mtl`` quite a bad idea.
+
+Pros:
+
+* Simple and familiar. It's easy to reason about effects when there's a 1-1
+  correspondence between classes and transformers. Virtually everyone is
+  familiar with ``mtl`` and can quickly learn to apply this approach.
+
+* Flexibility. Given a monad ``m``, one can easily add an effect to it by
+  putting the relevant transformer on top. It's convenient to use effects
+  locally, not only application-wide.
+
+Cons:
+
+* Bad run-time performance. Having a transformer per effect means no flattening.
+  This boils down to ``ReaderT (a, b, c, d) m`` being more performant than
+  ``ReaderT a (ReaderT b (ReaderT c (ReaderT d m)))`` (same goes for ``StateT``,
+  ``WriterT``, ``ExceptT``, etc). Since transformers for many domain-specific
+  effects are isomorphic to one of these general purpose effects, having a dozen
+  of nested transformers isn't such a good idea. (@int-index has benchmarks to back up
```

Comment:
Maybe add reference to benchmarks is these benchmarks are somewhere in the gist or github?

Reply:
https://github.com/int-index/ether/blob/master/bench/Bench.hs